src/hotspot/share/oops/symbolHandle.hpp

```diff
@@ -43,16 +44,29 @@
 // probe() and lookup_only() will increment the refcount if symbol is found.
 template <bool TEMP>
 class SymbolHandleBase : public StackObj {
+  static Symbol* volatile _cleanup_delay_queue[];
+  static volatile uint _cleanup_delay_index;
```

Comment:
Putting the delay queue implementation in SymbolHandleBase<> makes unnecessary and unused
data and possibly copies of the code. It is only used in the case where the template parameter is true.
Better would be to put the cleanup delay queue in a separate, non-templated, class. The entire
implementation of the queue could then be in the .cpp file. (I suggest the overhead of an out-of-line
call to add to the queue is in the noise, given that adding to the queue performs 3 atomic RMW
operations.)  So something like this:

```
// In symbolHandle.hpp.
class TempSymbolCleanupDelayer : AllStatic {
  // Or make these file-scoped statics in the .cpp file.
  static const uint QueueSize = 128;
  static Symbol* volatile _queue[];
  static volatile uint _index;

public:
  static void delay_cleanup(Symbol* s);
};

// In symbolHandle.cpp.
Symbol* volatile TempSymbolCleanupDelayer::_queue[QueueSize] = {};
volatile uint TempSymbolCleanupDelayer::_index = 0;

void TempSymbolCleanupDelayer::delay_cleanup(Symbol* sym) {
  assert(sym != nullptr, "precondition");
  sym->increment_refcount();
  uint i = Atomic::add(&_index, 1u) % QueueSize;
  Symbol* old = Atomic::xchg(&_queue, sym);
  Symbol::maybe_decrement_refcount(old);
}
```

Code is completely untested.  It incorporates suggestions I'm making elsewhere in this review too.

Reply:
This seems like a reasonable suggestion.  It would be better in as a singleton in the cpp file.